File: fortran90/day10_part1_2019.f90

program asteroid_visibility
  implicit none
  integer, parameter :: max_size = 500
  logical :: asteroids(max_size, max_size)
  integer :: num_rows, num_cols, max_count, i, j
  
  call read_asteroids('input.txt', asteroids, num_rows, num_cols)
  max_count = find_best_asteroid_location(asteroids, num_rows, num_cols)
  print *, max_count
  
contains

  subroutine read_asteroids(filename, asteroids, num_rows, num_cols)
    implicit none
    character(len=*), intent(in) :: filename
    logical, intent(out) :: asteroids(:,:)
    integer, intent(out) :: num_rows, num_cols
    integer :: i, j
    character(len=max_size) :: line
    integer :: ios
    open(unit=10, file=filename, status='old', action='read', iostat=ios)
    if (ios /= 0) then
      print *, 'Error opening file'
      stop
    end if
    num_rows = 0
    do
      read(10, '(A)', iostat=ios) line
      if (ios /= 0) exit
      num_rows = num_rows + 1
      num_cols = len_trim(line)
      do j = 1, num_cols
        asteroids(num_rows, j) = line(j:j) == '#'
      end do
    end do
    close(10)
  end subroutine read_asteroids

  function find_best_asteroid_location(asteroids, num_rows, num_cols) result(max_count)
    implicit none
    logical, intent(in) :: asteroids(:,:)
    integer, intent(in) :: num_rows, num_cols
    integer :: max_count, count, i, j
    max_count = 0
    do i = 1, num_rows
      do j = 1, num_cols
        if (asteroids(i, j)) then
          count = count_visible_asteroids(asteroids, j, i, num_rows, num_cols)
          max_count = max(max_count, count)
        end if
      end do
    end do
  end function find_best_asteroid_location

  function count_visible_asteroids(asteroids, x, y, num_rows, num_cols) result(count)
    implicit none
    logical, intent(in) :: asteroids(:,:)
    integer, intent(in) :: x, y, num_rows, num_cols
    integer :: count, other_x, other_y, dx, dy, g
    real :: angle
    logical :: seen(0:36000)
    count = 0
    seen = .false.
    do other_y = 1, num_rows
      do other_x = 1, num_cols
        if (asteroids(other_y, other_x) .and. (other_x /= x .or. other_y /= y)) then
          dx = other_x - x
          dy = other_y - y
          g = gcd(dx, dy)
          dx = dx / g
          dy = dy / g
          angle = atan2(real(dy), real(dx))
          angle = angle * 10000.0
          if (angle < 0) angle = angle + 36000.0
          if (.not. seen(int(angle))) then
            count = count + 1
            seen(int(angle)) = .true.
          end if
        end if
      end do
    end do
  end function count_visible_asteroids

  recursive function gcd(a, b) result(res)
    implicit none
    integer, intent(in) :: a, b
    integer :: res
    if (b == 0) then
      res = abs(a)
    else
      res = gcd(b, mod(a, b))
    end if
  end function gcd

end program asteroid_visibility

File: c/day10_part1_2019.c

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>

typedef struct {
    int x, y;
} Point;

int gcd(int a, int b) {
    while (b) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

int countVisibleAsteroids(bool **asteroids, int width, int height, int x, int y) {
    int count = 0;
    bool seen[2000][2000] = {false};
    for (int otherY = 0; otherY < height; otherY++) {
        for (int otherX = 0; otherX < width; otherX++) {
            if (asteroids[otherY][otherX] && !(otherX == x && otherY == y)) {
                int dx = otherX - x;
                int dy = otherY - y;
                int g = gcd(abs(dx), abs(dy));
                int sx = dx / g;
                int sy = dy / g;
                if(!seen[sx+1000][sy+1000]) {
                    seen[sx+1000][sy+1000]=true;
                    count++;
                }
            }
        }
    }
    return count;
}

int findBestAsteroidLocation(bool **asteroids, int width, int height) {
    int maxCount = 0;
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            if (asteroids[y][x]) {
                int count = countVisibleAsteroids(asteroids, width, height, x, y);
                if (count > maxCount) {
                    maxCount = count;
                }
            }
        }
    }
    return maxCount;
}

int main() {
    FILE *file = fopen("input.txt", "r");
    if (file == NULL) {
        perror("Error opening file");
        return 1;
    }

    char line[1024];
    int height = 0;
    int width = 0;
    while (fgets(line, sizeof(line), file) != NULL) {
        if(width==0){
            for (width = 0; line[width]!='\0' && line[width]!='\n'; width++);
        }
        height++;
    }

    fseek(file, 0, SEEK_SET);
    
    bool **asteroids = (bool **)malloc(height * sizeof(bool *));
    for (int i = 0; i < height; i++) {
        asteroids[i] = (bool *)malloc(width * sizeof(bool));
    }

    int row = 0;
    while (fgets(line, sizeof(line), file) != NULL) {
        for (int col = 0; col < width; col++) {
            asteroids[row][col] = line[col] == '#';
        }
        row++;
    }
    fclose(file);
    

    int maxCount = findBestAsteroidLocation(asteroids, width, height);
    printf("%d\n", maxCount);
    
    for (int i = 0; i < height; i++) {
        free(asteroids[i]);
    }
    free(asteroids);
    
    return 0;
}
